### PR TITLE
Add failsafe endstop while homing

### PR DESCRIPTION
**Pull request by Speaka**
### Description
First quick and dirty attempt to get failsafe endstops. Related to the mentioned FR, homing bump checks for inactive endstop after first hit and bump. If no change is detected, machine is stopped.
 I'm no coding pro, i copy paste the parts i understand until they work. So any help and improvements are appreciated. If I'm going the wrong way to contribute this, let me know.

_What works:_
- Checks related axis and depending endstops (min/max) depending on homedir.
- Only XYZ on kartesian, idk delta

_What needs to be done:_
- clean code since i dont know marlin structure so well to get the optimal and intendend way to integrate the checks.
- maybe optimize the used functions to reduce redundant code

### Benefits
Checking for Endstops state after first hit and moving back, makes sure to detect failed endstops (broken or stucked switch or cable) while homing.
Basically it detects the rising (falling) flank of the es signal to be sure both states are possible and nothing is broken

### Related Issues
#17895

### CODE REVIEW

**Review by Speaka — COMMENTED** [review]
updated the changes

**Comment by Speaka** [comment]
Shrunk the code. Testing and improvements are highly welcome.

**Comment by ghost** [comment]
Bear in mind, in some cases homing_bump is set to "0" (for "sensorless homing" on TMC drivers) so you might need to add some sanity checks.

**Comment by Speaka** [comment]
> Bear in mind, in some cases homing_bump is set to "0" (for "sensorless homing" on TMC drivers) so you might need to add some sanity checks.

Won't this be caught by line 1592: `  if (bump) {`?
My changes are inside of this if statement.

**Comment by thinkyhead** [comment]
I see that this will deal with the case of an endstop which is stuck ON, but for those which are stuck in the OFF state there will still be axis grinding. At least it will abort before grinding more than one axis.
